kong/cmd/utils/prefix_handler.lua

```diff
@@ -73,78 +73,85 @@ local function gen_default_ssl_cert(kong_config, target)
+      else
+        key = openssl_pkey.new { bits = 2048 }
+      end
 
-  if not pl_path.exists(ssl_cert) and not pl_path.exists(ssl_cert_key) then
-    log.verbose("generating %s SSL certificate and key", target or "default")
-
-    local key = openssl_pkey.new { bits = 2048 }
-
-    local crt = x509.new()
-    assert(crt:set_pubkey(key))
-    assert(crt:set_version(3))
-    assert(crt:set_serial_number(openssl_bignum.from_binary(openssl_rand.bytes(16))))
-
-    -- last for 20 years
-    local now = os.time()
-    assert(crt:set_not_before(now))
-    assert(crt:set_not_after(now + 86400 * 20 * 365))
-
-    local name = assert(x509_name.new()
-      :add("C", "US")
-      :add("ST", "California")
-      :add("L", "San Francisco")
-      :add("O", "Kong")
-      :add("OU", "IT Department")
-      :add("CN", "localhost"))
-
-    assert(crt:set_subject_name(name))
-    assert(crt:set_issuer_name(name))
-
-    -- Not a CA
-    assert(crt:set_basic_constraints { CA = false })
-    assert(crt:set_basic_constraints_critical(true))
-
-    -- Only allowed to be used for TLS connections (client or server)
-    assert(crt:add_extension(x509_extension.new("extendedKeyUsage",
-                                                "serverAuth,clientAuth")))
-
-    -- RFC-3280 4.2.1.2
-    assert(crt:add_extension(x509_extension.new("subjectKeyIdentifier", "hash", {
-      subject = crt
-    })))
-
-    -- All done; sign
-    assert(crt:sign(key))
-
-    do -- write key out
-      local fd = assert(io.open(ssl_cert_key, "w+b"))
-      local pem = assert(key:to_PEM("private"))
-      assert(fd:write(pem))
-      fd:close()
-    end
+      local crt = x509.new()
+      assert(crt:set_pubkey(key))
+      assert(crt:set_version(3))
+      assert(crt:set_serial_number(openssl_bignum.from_binary(openssl_rand.bytes(16))))
+
+      -- last for 20 years
+      local now = os.time()
+      assert(crt:set_not_before(now))
+      assert(crt:set_not_after(now + 86400 * 20 * 365))
+
+      local name = assert(x509_name.new()
+        :add("C", "US")
+        :add("ST", "California")
+        :add("L", "San Francisco")
+        :add("O", "Kong")
+        :add("OU", "IT Department")
+        :add("CN", "localhost"))
+
+      assert(crt:set_subject_name(name))
+      assert(crt:set_issuer_name(name))
+
+      -- Not a CA
+      assert(crt:set_basic_constraints { CA = false })
+      assert(crt:set_basic_constraints_critical(true))
+
+      -- Only allowed to be used for TLS connections (client or server)
+      assert(crt:add_extension(x509_extension.new("extendedKeyUsage",
+                                                  "serverAuth,clientAuth")))
+
+      -- RFC-3280 4.2.1.2
+      assert(crt:add_extension(x509_extension.new("subjectKeyIdentifier", "hash", {
+        subject = crt
+      })))
+
+      -- All done; sign
+      assert(crt:sign(key))
+
+      do -- write key out
+        local fd = assert(io.open(ssl_cert_key, "w+b"))
```

Comment:
One thing we should really be doing, is to protect the private keys better. In Hybrid mode we use the `open` system call to create the config cache file in `0600` mode if one does not exist yet. I think we should do the same for the private key here.

The old code wasn't doing it, I just realized that after reading the PR.

Example:
https://github.com/Kong/kong/blob/next/kong/clustering.lua#L668-L671

Reply:
I added additional commit to create private keys with 0600.